ops/model.py

```diff
@@ -176,20 +245,26 @@ def get_binding(self, binding_key: typing.Union[str, 'Relation']) -> 'Binding':
         return self._bindings.get(binding_key)
 
 
-class _ModelCache:
+_T = TypeVar('_T', bound='UnitOrApplication')
+
 
+class _ModelCache:
     def __init__(self, meta: 'ops.charm.CharmMeta', backend: '_ModelBackend'):
+        if typing.TYPE_CHECKING:
+            _weakcachetype = weakref.WeakValueDictionary[
+                Tuple[Any, ...], Optional[UnitOrApplication]]
+
         self._meta = meta
         self._backend = backend
-        self._weakrefs = weakref.WeakValueDictionary()
+        self._weakrefs = weakref.WeakValueDictionary()  # type: _weakcachetype
 
-    def get(self, entity_type, *args):
+    def get(self, entity_type: Type['_T'], *args: Any) -> '_T':
         key = (entity_type,) + args
         entity = self._weakrefs.get(key)
         if entity is None:
             entity = entity_type(*args, meta=self._meta, backend=self._backend, cache=self)
             self._weakrefs[key] = entity
-        return entity
+        return entity  # type: ignore
```

Comment:
since `_weakrefs` maps to (optional) `UnitOrApplication`, if we call `get(unit, 'foo')` the checker can't be sure whether `entity` is of type Unit or of type Application.

Reply:
Found a better way with overloads
![image](https://user-images.githubusercontent.com/6230162/169785781-2e85c279-ffb9-4219-8608-72d54b96b361.png)

Reply:
cool